### DISCUSSION TITLE
network/netfilter

### DISCUSSION THREAD

**Discussion by giscus** [isno/theByteBook#172] [General]
# network/netfilter

构建大规模高可用的分布式系统

https://www.thebyte.com.cn/network/netfilter.html

**Comment by yankaiqian** [comment]
“然后再依次经过 INPUT 链的 mangle、filter、nat表”，这里和图中不一致，建议改为： “然后再依次经过 INPUT 链的 mangle、filter表”。
或者说默认NAT表应该不包括 INPUT链。

**Reply by isno** [reply]
《Packet flow in Netfilter and General Networking》 这个图里面 INPUT 经过 nat 表。 刚才在服务器里面查了一下，也有NAT 表。这一节的配图我画错了，我修改一下。

**Comment by YellowPure** [comment]
这里 INPUT hook 和 OUTPUT hook 的箭头顺序是不是反了呢？https://www.netfilter.org/documentation/HOWTO/netfilter-hacking-HOWTO-3.html
NF_IP_LOCAL_IN 是发送到设备之前触发的 hook，那么是不是箭头应该是从 Application => Device 方向。同理，OUTPUT hook 应该是 Device => Application 方向。

**Comment by joy717** [comment]
之前也到这边看了佬写的这本书，感觉这块好像改过了。
这次看到这张图，眼前一亮。之前的一些困惑也顿时清晰了许多.
这些hook原来是在不同的协议层触发的，而在不同layer，因此处理的内容会有所偏重.

之前看的各种经典图，都没看到提及在什么layer触发什么hook。
wiki的这张图倒是有这个信息，但是内容太多，一直没能理解这边的协议层级信息。
![image](https://en.wikipedia.org/wiki/Netfilter#/media/File:Netfilter-packet-flow.svg)

如果看不到图片，可以点这边：https://en.wikipedia.org/wiki/Netfilter#/media/File:Netfilter-packet-flow.svg

特意登录，留个言，感谢
